tsl/src/continuous_aggs/refresh.c

```diff
@@ -532,50 +543,44 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 }
 
 /*
- * Refresh all continuous aggregates on a hypertable.
+ * Refresh a continuous aggregate on a chunk.
  *
  * The refreshing happens in a single transaction. For this to work correctly,
  * there must be no new invalidations written in the refreshed region during
- * the refresh. Therefore, the caller is responsible for proper locking to
- * ensure there are no invalidations (INSERTs, DELETEs, etc.). For instance,
- * exclusively locking the hypertable or the individual chunks covered by the
- * region would work.
+ * the refresh. Therefore, it locks exclusively the chunk to
+ * ensure there are no invalidations (INSERTs, DELETEs, etc.).
  */
-void
-continuous_agg_refresh_all(const Hypertable *const ht, const int64 start, const int64 end,
-						   const int32 chunk_id)
+Datum
+continuous_agg_refresh_chunk(PG_FUNCTION_ARGS)
 {
+	Oid cagg_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	ContinuousAgg *cagg = get_cagg_by_relid(cagg_relid);
+	Chunk *chunk = ts_chunk_get_by_relid(chunk_relid, true);
 	Catalog *catalog = ts_catalog_get();
-	List *caggs = ts_continuous_aggs_find_by_raw_table_id(ht->fd.id);
-	Dimension *dim = hyperspace_get_open_dimension(ht->space, 0);
 	InternalTimeRange refresh_window = {
-		.type = ts_dimension_get_partition_type(dim),
-		.start = start,
-		.end = end,
+		.type = cagg_get_partition_type(cagg),
+		.start = ts_chunk_primary_dimension_start(chunk),
+		.end = ts_chunk_primary_dimension_end(chunk),
 	};
-	ListCell *lc;
 
-	/* We're not doing any specific permissions checks here. It's assumed that
-	 * whoever calls this function has done appropriate checks for the
-	 * operation. For instance, if this is called as a result of
-	 * "refresh-on-drop", it is assumed that refresh can happen if the user is
-	 * permitted to drop data. */
+	/* Like regular materialized views, require owner to refresh. */
+	if (!pg_class_ownercheck(cagg->relid, GetUserId()))
+		aclcheck_error(ACLCHECK_NOT_OWNER,
+					   get_relkind_objtype(get_rel_relkind(cagg->relid)),
+					   get_rel_name(cagg->relid));
+
+	PreventCommandIfReadOnly(REFRESH_FUNCTION_NAME);
```

Comment:
There's a macro in `compat.h` to handle this now: `TS_PREVENT_FUNC_IF_READ_ONLY()`